driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java

```diff
@@ -173,6 +181,34 @@ public Boolean getRetryWrites() {
         return retryWrites;
     }
 
+    private <R> Supplier<R> decorateWriteWithRetries(final RetryState retryState, final Supplier<R> writeFunction) {
+        return new RetryingSyncSupplier<>(retryState, CommandOperationHelper::chooseAndMutateRetryableWriteException,
+                this::shouldAttemptToRetryWrite, writeFunction);
+    }
+
+    private <R> AsyncCallbackSupplier<R> decorateWriteWithRetries(final RetryState retryState,
+            final AsyncCallbackSupplier<R> writeFunction) {
+        return new RetryingAsyncCallbackSupplier<>(retryState, CommandOperationHelper::chooseAndMutateRetryableWriteException,
+                this::shouldAttemptToRetryWrite, writeFunction);
+    }
+
+    private boolean shouldAttemptToRetryWrite(final RetryState retryState, final Throwable attemptFailure) {
+        BulkWriteTracker bulkWriteTracker = retryState.attachment(AttachmentKeys.bulkWriteTracker()).orElseThrow(Assertions::fail);
+        /* A retry predicate is called only if there is at least one more attempt left. Here we maintain attempt counters manually
+         * and emulate the above contract by returning `false` at the very beginning of the retry predicate. */
+        if (bulkWriteTracker.lastAttempt()) {
+            return false;
+        }
+        boolean decision = CommandOperationHelper.shouldAttemptToRetryWrite(retryState, attemptFailure);
```

Comment:
Nit: statically import this method

Reply:
Can't do because there is a method with the same signature in `MixedBulkWriteOperation`.